springboard_iogt/views.py

```diff
@@ -0,0 +1,70 @@
+import random
+from datetime import datetime
+from itertools import chain
+
+from pyramid.view import view_config
+from elasticutils import F
+
+from springboard.views.base import SpringboardViews
+
+from springboard_iogt.utils import randomize_query
+
+
+class IoGTViews(SpringboardViews):
+
+    @view_config(route_name='home',
+                 renderer='springboard_iogt:templates/home.jinja2')
+    def index_view(self):
+        return self.context(recent_content=self.recent_content())
+
+    def recent_content(self):
+        # get 2 most recent pages and their categories
+        [page1, page2] = self.all_pages.filter(
+            language=self.language).order_by('-created_at')[:2]
+        categories = self.all_categories.filter(
+            uuid__in=filter(lambda uuid: uuid is not None,
+                            [page1.primary_category, page2.primary_category]))
+        categories = dict((category.uuid, category) for category in categories)
+        category1 = (categories[page1.primary_category]
+                     if page1.primary_category else None)
+        category2 = (categories[page2.primary_category]
+                     if page2.primary_category else None)
+
+        # random seed that changes hourly
+        seed = datetime.utcnow().replace(
+            minute=0, second=0, microsecond=0)
+        seed = (seed - datetime.utcfromtimestamp(0)).total_seconds()
+        seed = int(seed)
+
+        # get random categories and exclude categories of 2 most recent pages
+        categories = self.all_categories.filter(
+            ~F(uuid__in=categories.keys()), language=self.language)
+        categories = randomize_query(categories, seed=seed)
+        # filter to exclude the 2 most recent pages
+        f_exclude_pages = ~F(uuid__in=[page1.uuid, page2.uuid])
+
+        def do_query(limit):
+            most_recent = [(category1, page1), (category2, page2)]
+            most_recent_per_category = []
+            # NOTE: this is bad if limit is large
+            # We are fetching pages individually, because we
+            # can't use facets or aggregates.
+            for category in categories[:limit - 2]:
+                try:
+                    [page] = self.all_pages.filter(
+                        f_exclude_pages,
+                        primary_category=category.uuid).order_by(
+                        '-created_at')[:1]
+                    most_recent_per_category.append((category, page))
+                except ValueError:
+                    pass
+
+            results = [(c.to_object() if c else None, p.to_object())
+                       for c, p
+                       in chain(most_recent, most_recent_per_category)]
+            # randomize position of most_recent content
+            random.seed(seed)
+            random.shuffle(results)
+            return results
+
+        return do_query
```

Comment:
Are we returning a function here to try & not evaluate this stuff unless called explicitly in the template?